### ISSUE TITLE
Camera Limit smoothing does not work when new limit does not completely bound existing camera position limits

### ISSUE THREAD

**Issue by Validifyed** [godotengine/godot#32596]
**Godot version:** 3.1/3.2a1

**OS/device including version:** Win10 x64

**Issue description:** 
When dynamically setting camera limits, if the limit is extended beyond current limits, the camera smoothing for the limit will slowly interpolate the camera to its new position. - This is the limit smoothing behavior working as expected.

When creating a new limit which does not fully contain the current camera view-port, the camera position snaps to its new position within the new limit, rather than smoothly interpolating to this new position.

**Steps to reproduce:**
Place camera at 0,0, enable limits at huge bounds (+/- 10000), enable limit smoothing.
Set script to change limit such that one of the X bounds bisects the default/unlimited camera position.
Camera will snap to new position in limit rather than smoothing to it.

**Minimal reproduction project:**

[Camera_Bug.zip](https://github.com/godotengine/godot/files/3694644/Camera_Bug.zip)

Camera starts limited, moving left will reach a border where the limits are massively extended (in effect - removed), moving back into the limited area will cause the camera to do the aforementioned snapping behavior due to limits being re-assigned.

**Comment by KoBeWi** [comment]
Still valid in 3.2.3

This is kind of expected though as the camera is supposed to never leave its bounds. Although `offset` is an exception, so the desired behavior could be simulated with a script.

**Comment by ekumlin** [comment]
This is still valid in 3.3.

From my perspective, the existing behavior is undesirable; camera bounding limits are adjustable dynamically via code, which could be (for example) something like Metroid where the camera is moving between rooms, but can still move freely within each room. In effect, you're just setting a new bounded area for that camera to move in, so it would make sense for smoothing to apply as it moves to those new bounds (i.e. to the next room).

If someone wanted this behavior to not occur, they could either set `smoothed = false` before adjusting the limits, or `reset_smoothing()` immediately after. By default I think it would be safe to assume that smoothing is desirable.

**Comment by Calinou** [comment]
> Edit: I think the change would be as simple as wrapping camera_2d.cpp:181-195 with if (!limit_smoothing_enabled) { ... }. I've never set up a local build for Godot but if this is a change we're willing to take, I could give it a shot.

Feel free to open a pull request for this against the `master` branch :slightly_smiling_face: 

The pull request can then be cherry-picked to the `3.x` branch.

**Comment by ekumlin** [comment]
Thanks Calinou. I was able to get Godot building locally (which was remarkably easy!) and make the change.

Couple things to note:

- I tested the fix on `3.x` and it works perfectly in my scenario.
- I made the change in `master` and created that PR above. We may run into merge conflicts cherry-picking into `3.x` because of the change in `Camera2D` from `MARIGN_*` to `SIDE_*` constants.
- Unfortunately, I couldn't test the change on the `master` branch directly because I could not get my project (or the minimal repro) to run on 4.0 because of some `TileMap` issues I ran into.

Let me know if there are any issues with the PR and I'll do my best to address them.
